gh-125862: Context decorator support for generators and async functions

agronholm opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-125862 -->
* Issue: gh-125862
<!-- /gh-issue-number -->

agronholm committed:
gh-125862: Improve context decorator support for generators and async functions

agronholm committed:
Merge branch 'main' into fix-issue-125862

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

agronholm committed:
Manually iterate coroutines to avoid asyncio use

agronholm committed:
Merge branch 'main' into fix-issue-125862

Zac-HD commented on Lib/contextlib.py:
```diff
@@ -79,11 +82,32 @@ def _recreate_cm(self):
+
+        def gen_inner(*args, **kwds):
+            with self._recreate_cm():
+                yield from func(*args, **kwds)
```
```suggestion
            with self._recreate_cm(), closing(func(*args, **kwds)) as gen:
                yield from gen
```

I think we should be using `with closing()` and `async with aclosing()` in these wrappers, to ensure that any resources held by the wrapped function are cleaned up promptly.  (see e.g. [PEP-533 motivation](https://peps.python.org/pep-0533/#background-and-motivation))

agronholm replied:
Done.

agronholm committed:
Merge branch 'main' into fix-issue-125862

agronholm committed:
Merge branch 'main' into fix-issue-125862

agronholm committed:
Make sure we at least try to close the generators

agronholm committed:
Merge remote-tracking branch 'fork/fix-issue-125862' into fix-issue-125862

agronholm committed:
Use (a)closing

agronholm committed:
Merge branch 'main' into fix-issue-125862

gpshead committed:
Merge remote-tracking branch 'origin/main' into fix-issue-125862

gpshead committed:
Use _private imports

gpshead committed:
performance: only define inner funcs on the branch that uses them, lazy import inspect

gpshead committed:
do not delete that newline (small diff)

gpshead committed:
reword reST news entry

gpshead committed:
Preserve generator return value; expand decorator test coverage

ContextDecorator's generator wrapper now uses ``return (yield from gen)``
so a wrapped generator's StopIteration value is not lost.

Tests: drop the unused run_test() helper, normalize quote style, and add
exception-path and early-close coverage for the new generator and async
generator wrappers so closing()/aclosing() behavior is exercised.

gpshead committed:
Document ContextDecorator generator/coroutine handling

Add versionchanged notes to ContextDecorator and AsyncContextDecorator
describing the new behavior when decorating generator, coroutine, and
async generator functions, and call out that asend()/athrow() are not
forwarded through the async generator wrapper.  Reword the
AsyncContextDecorator intro now that synchronous callables are accepted.

gpshead committed:
Add What's New entry for ContextDecorator generator support

gpshead committed:
simplify lazy import syntax

gpshead committed:
Reword AsyncContextDecorator docs to lead with intended async use

State the primary use (coroutine and async generator functions) up front
and move the synchronous-input behavior into a note that points users at
ContextDecorator when the calling convention change is not wanted.

gpshead committed:
Reword test comments to state the contract, not the mechanism

The early-stop tests asserted ordering with comments that named closing()
and aclosing() as the reason; restate as the invariant being tested so
the comments stay correct if the implementation changes.

gpshead committed:
Test send/throw forwarding for the sync wrapper; pin async limitation

The sync generator wrapper uses ``yield from`` so .send() and .throw()
reach the wrapped generator; lock that in.  The async generator wrapper
re-yields with ``async for`` so asend()/athrow() do not reach the
wrapped generator; pin that documented behavior so a future change that
starts forwarding will fail and prompt a doc update.

gpshead committed:
also add my name in whatsnew
